telegram/_messageentity.py

```diff
@@ -140,6 +140,92 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
 
         return super().de_json(data=data, bot=bot)
 
+    @staticmethod
+    def adjust_message_entities_to_utf_16(
+        text: str, entities: Sequence["MessageEntity"]
+    ) -> Sequence["MessageEntity"]:
+        """Utility functionality for converting the offset and length of entities from
+        Unicode (`str`) to UTF-16 (`utf-16-le` encoded `bytes`).
+
+        Tip:
+        Only the offsets and lengths calulated in UTF-16 is acceptable by the Telegram Bot API.
+        If they are calculated using the Unicode string (`str` object), errors may occur
+        when the text contains characters that are not in the Basic Multilingual Plane (BMP).
+
+        Examples:
+        Below is a snippet of code that demonstrates how to use this function to convert
+        entities from Unicode to UTF-16 space. The `unicode_entities` are calculated in
+        Unicode and the `utf_16_entities` are calculated in UTF-16.
+
+        ```python
+        text = "𠌕 bold 𝄢 italic underlined: 𝛙𝌢𑁍"
+        #         ^^^^   ^^^^^^             ^^^
+        unicode_entities = [
+            MessageEntity(offset=2, length=4, type=MessageEntity.BOLD),
+            MessageEntity(offset=9, length=6, type=MessageEntity.ITALIC),
+            MessageEntity(offset=28, length=3, type=MessageEntity.UNDERLINE),
+        ]
+        utf_16_entities = MessageEntity.adjust_message_entities_to_utf_16(
+            text, unicode_entities
+        )
+        await bot.send_message(
+            chat_id=123,
+            text=text,
+            entities=utf_16_entities,
+        )
+        # utf_16_entities[0]: offset=3, length=4
+        # utf_16_entities[1]: offset=11, length=6
+        # utf_16_entities[2]: offset=30, length=6
+        ```
+
+        Args:
+            text (:obj:`str`): The text that the entities belong to
+            entities (List[:class:`telegram.MessageEntity`]): Sequence of entities
+                with offset and length calculated in Unicode
+
+        Returns:
+            List[:class:`telegram.MessageEntity`]: Sequence of entities with offset and length
+                calculated in UTF-16 encoding
+        """
+        offsets_set = set()
+        # collect all offsets occured
+        for entity in entities:
+            offsets_set.add(entity.offset)
+            offsets_set.add(entity.offset + entity.length)
+        # get sorted offsets
+        offsets_list = sorted(offsets_set)
+        # indexes_map stores the index of each `cut_offset` in `offsets_list`
+        indexes_map = {cut_offset: i for i, cut_offset in enumerate(offsets_list)}
+        # calculate the length of each slice in utf-16
+        accumulated_length = 0
+        utf16_lengths = []
+        # what utf16_lengths stores:
+        # utf16_lengths[indexes_map[offset]] = len(text[:offset].encode("utf-16-le")) // 2
```

Comment:
this comment is not particularly helpful. A code comment should explain what the meaning of the code is or why it is done. In this case, the meaning of the code would be something like `for each split_mark in unicode-space it stores the corresponding split_mark in utf_16-space`